src/plugins/layout/shadowdom.ts

```diff
@@ -272,25 +208,93 @@ export class ShadowDom {
       this.removeAllClasses(next);
     }
 
-    let removed = this.removedNodes.childNodes;
-    for (let i = 0; i < removed.length; i++) {
-      let next = removed[i] as IShadowDomNode;
-      if (!this.hasClass(next, NewNodeClassName)) {
-        summary.removedNodes.push(next);
-      }
+    let removedNodes = this.removedNodes.childNodes;
+    for (let i = 0; i < removedNodes.length; i++) {
+      traverseNodeTree(removedNodes[i], (shadowNode: IShadowDomNode) => {
+        if (this.hasClass(shadowNode, NewNodeClassName)) {
+          delete shadowNode.node[NodeIndex];
+        } else if (this.hasClass(shadowNode, MovedNodeClassName)) {
+          summary.removedNodes.push(shadowNode);
+        }
+      });
     }
 
     return summary;
   }
 
+  public createDomIndexJson(): object {
+
+    function writeIndex(node: Node, json: object) {
+      let index = getNodeIndex(node);
+      let childJson = {};
+      let nextChild = node.firstChild;
+      json[index] = nextChild ? childJson : null;
+      while (nextChild) {
+        writeIndex(nextChild, childJson);
+        nextChild = nextChild.nextSibling;
+      }
+    }
+
+    let indexJson = {};
+    writeIndex(document, indexJson);
+
+    return indexJson;
+  }
+
+  public createShadowDomIndexJson(): object {
```

Comment:
Done, merged two functions in one with a help of an extra argument

---
In reply to: [131239745](https://github.com/Microsoft/clarity-js/pull/25#discussion_r131239745) [](ancestors = 131239745)